motionplan/motionPlanner.go

```diff
@@ -464,104 +463,74 @@ func CheckPlan(
-		return errors.New("plan must have at least one element")
-	}
-
-	// construct solverFrame
-	// Note that this requires all frames which move as part of the plan, to have an
-	// entry in the very first plan waypoint
-	sf, err := newSolverFrame(fs, checkFrame.Name(), frame.World, plan[0])
+	sf, err := newSolverFrame(fs, checkFrame.Name(), frame.World, currentInputs)
 	if err != nil {
 		return err
 	}
 
-	// construct planager
 	sfPlanner, err := newPlanManager(sf, fs, logger, defaultRandomSeed)
 	if err != nil {
 		return err
 	}
 
-	// convert plan into nodes
-	planNodes, err := sf.planToNodes(plan)
-	if err != nil {
-		return err
-	}
-
 	// This should be done for any plan whose configurations are specified in relative terms rather than absolute ones.
 	// Currently this is only TP-space, so we check if the PTG length is >0.
 	// The solver frame will have had its PTGs filled in the newPlanManager() call, if applicable.
 	relative := len(sf.PTGSolvers()) > 0
 
-	if relative {
-		// get pose of robot along the current trajectory it is executing
-		lastPose, err := sf.Transform(currentInputs)
-		if err != nil {
-			return err
-		}
-
-		// where ought the robot be on the plan
-		pathPosition := spatialmath.PoseBetweenInverse(errorState, currentPosition)
-
-		// absolute pose of the previous node we've passed
-		formerRunningPose := spatialmath.PoseBetweenInverse(lastPose, pathPosition)
-
-		// convert planNode's poses to be in absolute coordinates
-		if planNodes, err = rectifyTPspacePath(planNodes, sf, formerRunningPose); err != nil {
-			return err
-		}
+	offsetPlan, err := OffsetPlan(plan, errorState)
+	if err != nil {
+		return err
 	}
 
-	// adjust planNodes by the errorState
-	// this only changes a node's pose and not its inputs
-	planNodes = transformNodes(planNodes, errorState)
-
-	// pre-pend node with current position of robot to planNodes
-	// Note that currentPosition is assumed to have accounted for the errorState
-	// Note that currentInputs is assumed to have NOT accounted for the errorState
-	planNodes = append([]node{&basicNode{pose: currentPosition, q: currentInputs}}, planNodes...)
-
 	// create constraints
 	if sfPlanner.planOpts, err = sfPlanner.plannerSetupFromMoveRequest(
-		currentPosition,                    // starting pose
-		planNodes[len(planNodes)-1].Pose(), // goalPose
-		plan[0],                            // starting configuration
+		currentPosition, // starting pose
+		offsetPlan.Path()[len(offsetPlan.Path())-1][checkFrame.Name()].Pose(), // goalPose
+		currentInputs,
 		worldState,
 		nil, // no pb.Constraints
 		nil, // no plannOpts
 	); err != nil {
 		return err
 	}
 
-	// go through plan and check that we can move from plan[i] to plan[i+1]
-	var totalTravelDistanceMM float64
-	for i := 0; i < len(planNodes)-1; i++ {
-		currentPose := planNodes[i].Pose()
-		nextPose := planNodes[i+1].Pose()
-		startConfiguration := planNodes[i].Q()
-		endConfiguration := planNodes[i+1].Q()
+	createSegment := func(
```

Comment:
Could this be a normal function rather than an anonymous function? At first blush, it seems like the only value its capturing (other than its parameters) is `relative`, which seems easy to make a parameter.

Reply:
I thought about this a little more and am not a huge fan of this.  Exposing it elsewhere will make it possible for others to mistakenly use this code, when IMO it should only be accessible within this function.  Eventually RSDK-5007 should be merged which will let us clean this up by calling the intended way to handle this `CheckStateConstraintsAcrossSegment`